internal/metrics/resource/resources.go

```diff
@@ -0,0 +1,96 @@
+package resource
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/rhobs/multicluster-observability-addon/internal/addon"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	addonapiv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+var (
+	mu          sync.Mutex
+	initialized bool
+	ErrNotOwner = fmt.Errorf("controller is not the owner")
+)
+
+func DeployDefaultResourcesOnce(ctx context.Context, c client.Client, logger logr.Logger, ns string) error {
```

Comment:
I would refactor this function in the following way:
- Determine what resources need to be deployed (currently IIUC the code is always deploying resources for both no-UWM & UWM)
- Depending on the use case try & get the root resources (PA for no-UWM & UWM) 
- If success we return
- Otherwise depending on the use case we call create

Also AFAIK mutex is not needed for these types of tasks

Reply:
I want to avoid triggering this resources creation on each reconcile of each managed cluster. My aim is to create it once at the start and not touch it again, unless it restarts. But this is a debatable approach.
I am limited by the interface provided by the addon framework. I don't know what is the trigger for the reconcile.
Given that, it is much simpler to generate it all once and avoid tracking what is updated in the collected signals.

Regarding the mutex, it is probably overkill. I did not know if managed clusters where reconciled concurrently (I assume this was my reasoning at the time). Got to check

Reply:
> They are not, its a queue of ManagedClusterAddOn so we are safe without it
Ack removing the mutex.

> I'm not sure I understand, the approach I proposed also does this but it also only deploys the resources that are strictly necessary & reacts if users accidentally deletes them.

I'm just worried it is duplicating the same processing, checking these common resources for each managed cluster, while it could do it only once. As a first approach I simplified things (like deploying all defaults), this is not necessarily the target implementation. With this I can do it on every managed cluster, I'm fine with it too.
As a target, I think that the best solution would be to have a dedicated controller for managing these default resources or create them from MCO. For example, I have some resource like the ScrapeConfigs defining what metrics to collect that are tightly coupled with the dashboards defined there, and I'll probably define them there.

And now that we will not reference all those default resources in the addon configs list, if the user changes them, no reconcile will be triggered when handling them here.

WDYT?